Web Inspector: Proposing feature: Display font selected as the font under identity in the font tab

francescorn opened the pull request:
#### 00dfac4ac7d775d4838023cd7dcf3072061ed7e1
<pre>
Web Inspector: Proposing feature: Display font selected as the font under identity in the font tab
<a href="https://bugs.webkit.org/show_bug.cgi?id=260177">https://bugs.webkit.org/show_bug.cgi?id=260177</a>

Reviewed by NOBODY (OOPS!).

The purpose of this pr is create All Fonts Properties Section that returns
all of the loaded fonts on the webpage consisting of system and custom.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getLoadedFonts):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.get usedFonts):
(WI.DOMNodeStyles.prototype.refresh):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype.initialLayout):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00dfac4ac7d775d4838023cd7dcf3072061ed7e1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52617 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52812 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/246 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51874 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26475 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40472 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51670 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26392 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42700 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21587 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23847 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43883 "Found 29 new test failures: inspector/css/add-css-property.html, inspector/css/css-property.html, inspector/css/forcePseudoState.html, inspector/css/generateFormattedText.html, inspector/css/getComputedPrimaryFontForNode.html, inspector/css/getMatchedStylesForNodeBackdropPseudoId.html, inspector/css/getMatchedStylesForNodeContainerGrouping.html, inspector/css/getMatchedStylesForNodeLargeSelectors.html, inspector/css/getMatchedStylesForNodeLayerGrouping.html, inspector/css/getMatchedStylesForNodeMarkerPseudoId.html ... (failure)") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7941 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/42885 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45761 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44398 "Found 29 new test failures: inspector/css/add-css-property.html, inspector/css/css-property.html, inspector/css/forcePseudoState.html, inspector/css/generateFormattedText.html, inspector/css/getComputedPrimaryFontForNode.html, inspector/css/getMatchedStylesForNodeBackdropPseudoId.html, inspector/css/getMatchedStylesForNodeContainerGrouping.html, inspector/css/getMatchedStylesForNodeLargeSelectors.html, inspector/css/getMatchedStylesForNodeLayerGrouping.html, inspector/css/getMatchedStylesForNodeMarkerPseudoId.html ... (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54389 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49065 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24655 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20842 "Found 27 new test failures: inspector/css/add-css-property.html, inspector/css/css-property.html, inspector/css/forcePseudoState.html, inspector/css/generateFormattedText.html, inspector/css/getComputedPrimaryFontForNode.html, inspector/css/getMatchedStylesForNodeBackdropPseudoId.html, inspector/css/getMatchedStylesForNodeContainerGrouping.html, inspector/css/getMatchedStylesForNodeLargeSelectors.html, inspector/css/getMatchedStylesForNodeLayerGrouping.html, inspector/css/getMatchedStylesForNodeMarkerPseudoId.html ... (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47840 "Found 14 new test failures: inspector/css/add-css-property.html, inspector/css/css-property.html, inspector/css/forcePseudoState.html, inspector/css/getMatchedStylesForNodeContainerGrouping.html, inspector/css/getMatchedStylesForNodeLargeSelectors.html, inspector/css/getMatchedStylesForNodeLayerGrouping.html, inspector/css/getMatchedStylesForNodeMarkerPseudoId.html, inspector/css/pseudo-element-matches-for-pseudo-element-node.html, inspector/css/pseudo-element-matches.html, inspector/css/resolve-variable-value.html ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25924 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42845 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46886 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/replaced-objects (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26768 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56552 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25648 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11625 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -903,6 +905,23 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::CSS::C
+Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent::getLoadedFonts()
+{
+    auto fontFamilyNames = JSON::ArrayOf<String>::create();
+    auto* domAgent = m_instrumentingAgents.persistentDOMAgent();
```
there's no guarantee that this always exists, so we need to check it before using it
```cpp
if (auto* domAgent = m_instrumentingAgents.persistentDOMAgent()) {
    for (auto* document : domAgent->documents()) {
        auto it = document->fonts()->createIterator(document->scriptExecutionContext());
        while (auto font = it.next()) {
            if (font->status() == FontFace::LoadStatus::Loaded)
                fontFamilyNames->addItem(font->family());
        }
    }
}
```

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -903,6 +905,23 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::CSS::C
+        auto iterator = document->fonts().get().createIterator(document->scriptExecutionContext());
+
+        while (auto font = iterator.next()) {
+            if (font->status() == FontFace::LoadStatus::Loaded)
```
is there a reason we have to wait for `FontFace::LoadStatus::Loaded`?  it seems like `family()` mainly looks at the CSS declaration to get a value

do we need to do anything for `FontFace::LoadStatus::Loading`?  i.e. what if the frontend calls this while some fonts are loading?  do we just drop those loading fonts entirely or would this get called again later?

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -122,7 +122,7 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
 
     // Public
-
+    get usedFonts() { return this._usedFonts; }
```
NIT: ideally this would be moved to match the order/location of the member variable

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +122,16 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
             this._fontVariationsGroup.rows = [emptyRow];
         }
+
+        console.log("In Font Details Panel");
+        console.log(this.nodeStyles.usedFonts);
```
oops

francescorn replied:
removed

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +122,16 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        for (let fontName of this.nodeStyles.usedFonts) {
+            let row = document.createElement("div");
+            row.textContent = fontName;
+            row.style.fontFamily = fontName;
```
this likely wont work as expected because the frontend is an entirely separate page from the backend, meaning that the `@font-family`/etc. to load the font wont exist

can you attach a screen shot of what this would look like?

francescorn replied:
My goal is add a fontName to each row and display each font in the font style.

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -462,6 +462,14 @@
         },
+        {
+            "name": "getLoadedFonts",
+            "description": "Returns all loaded fonts (system and custom).",
```
we already have `CSS.getSupportedSystemFontFamilyNames`.  if this also includes system font family names, perhaps we should remove that in favor of this?  or maybe just make this load custom fonts (i.e. `CSS.getLoadedCustomFonts`)?

francescorn replied:
Is `CSS.getSupportedSystemFontFamilyNames` in use in the frontend? I think we can keep `getLoadedFonts`, it is being called in other areas in the backend.

francescorn replied:
I see `CSS.getSupportedSystemFontFamilyNames` is in use in `CSSManager.js`. I was originally planning to support showing all the system and custom fonts in use.

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.h:
```diff
@@ -112,6 +112,7 @@ class InspectorCSSAgent final : public InspectorAgentBase , public Inspector::CS
     Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::CSS::CSSRule>> addRule(const Inspector::Protocol::CSS::StyleSheetId&, const String& selector);
     Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::CSS::CSSPropertyInfo>>> getSupportedCSSProperties();
     Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> getSupportedSystemFontFamilyNames();
+    Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> getLoadedFonts();
```
is this ever actually called by the frontend?

dcrousso reviewed (changes requested):
this seems like a good start, but i think you're missing some plumbing logic and have a few minor (possible) bugs

francescorn committed:
Web Inspector: Proposing feature: Display font selected as the font under identity in the font tab
https://bugs.webkit.org/show_bug.cgi?id=260177

Reviewed by NOBODY (OOPS!).

The purpose of this pr is create All Fonts Properties Section that returns
all of the loaded fonts on the webpage consisting of system and custom.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getLoadedFonts):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.get usedFonts):
(WI.DOMNodeStyles.prototype.refresh):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype.initialLayout):
